File: frameworks/native/appkit/ability_runtime/vpn_extension_context.cpp
/*
 * Copyright (c) 2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "vpn_extension_context.h"

#include "ability_connection.h"
#include "ability_manager_client.h"
#include "hilog_wrapper.h"
#include "hitrace_meter.h"

namespace OHOS {
namespace AbilityRuntime {
const size_t VpnExtensionContext::CONTEXT_TYPE_ID(std::hash<const char*> {} ("VpnExtensionContext"));
const std::string START_ABILITY_TYPE = "ABILITY_INNER_START_WITH_ACCOUNT";

int32_t VpnExtensionContext::ILLEGAL_REQUEST_CODE(-1);

ErrCode VpnExtensionContext::StartAbility(const AAFwk::Want &want) const
{
    HITRACE_METER_NAME(HITRACE_TAG_ABILITY_MANAGER, __PRETTY_FUNCTION__);
    HILOG_DEBUG("Start ability begin, ability:%{public}s.", want.GetElement().GetAbilityName().c_str());
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StartAbility(want, token_, ILLEGAL_REQUEST_CODE);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartAbility is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StartAbility(const AAFwk::Want &want, const AAFwk::StartOptions &startOptions) const
{
    HITRACE_METER_NAME(HITRACE_TAG_ABILITY_MANAGER, __PRETTY_FUNCTION__);
    HILOG_DEBUG("Start ability begin, ability:%{public}s.", want.GetElement().GetAbilityName().c_str());
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StartAbility(want, startOptions, token_,
        ILLEGAL_REQUEST_CODE);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartAbility is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StartAbilityAsCaller(const AAFwk::Want &want) const
{
    HITRACE_METER_NAME(HITRACE_TAG_ABILITY_MANAGER, __PRETTY_FUNCTION__);
    HILOG_DEBUG("Start ability as caller begin, ability:%{public}s.", want.GetElement().GetAbilityName().c_str());
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->
    StartAbilityAsCaller(want, token_, nullptr, ILLEGAL_REQUEST_CODE);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartAbilityAsCaller is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StartAbilityAsCaller(const AAFwk::Want &want,
    const AAFwk::StartOptions &startOptions) const
{
    HITRACE_METER_NAME(HITRACE_TAG_ABILITY_MANAGER, __PRETTY_FUNCTION__);
    HILOG_DEBUG("Start ability as caller begin, ability:%{public}s.", want.GetElement().GetAbilityName().c_str());
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StartAbilityAsCaller(want, startOptions, token_, nullptr,
        ILLEGAL_REQUEST_CODE);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartAbilityAsCaller is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StartAbilityByCall(
    const AAFwk::Want& want, const std::shared_ptr<CallerCallBack> &callback, int32_t accountId)
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    if (localCallContainer_ == nullptr) {
        localCallContainer_ = std::make_shared<LocalCallContainer>();
        if (localCallContainer_ == nullptr) {
            HILOG_ERROR("%{public}s failed, localCallContainer_ is nullptr.", __func__);
            return ERR_INVALID_VALUE;
        }
    }
    return localCallContainer_->StartAbilityByCallInner(want, callback, token_, accountId);
}

ErrCode VpnExtensionContext::ReleaseCall(const std::shared_ptr<CallerCallBack> &callback) const
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    if (localCallContainer_ == nullptr) {
        HILOG_ERROR("%{public}s failed, localCallContainer_ is nullptr.", __func__);
        return ERR_INVALID_VALUE;
    }
    return localCallContainer_->ReleaseCall(callback);
}

void VpnExtensionContext::ClearFailedCallConnection(const std::shared_ptr<CallerCallBack> &callback) const
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    if (localCallContainer_ == nullptr) {
        HILOG_ERROR("%{public}s failed, localCallContainer_ is nullptr.", __func__);
        return;
    }
    localCallContainer_->ClearFailedCallConnection(callback);
}

ErrCode VpnExtensionContext::ConnectAbility(
    const AAFwk::Want &want, const sptr<AbilityConnectCallback> &connectCallback) const
{
    HILOG_INFO("Connect ability begin, ability:%{public}s.", want.GetElement().GetAbilityName().c_str());
    ErrCode ret =
        ConnectionManager::GetInstance().ConnectAbility(token_, want, connectCallback);
    HILOG_INFO("VpnExtensionContext::ConnectAbility ErrorCode = %{public}d", ret);
    return ret;
}

ErrCode VpnExtensionContext::StartAbilityWithAccount(const AAFwk::Want &want, int32_t accountId) const
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    HILOG_INFO("%{public}d accountId:", accountId);
    (const_cast<Want &>(want)).SetParam(START_ABILITY_TYPE, true);
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StartAbility(
        want, token_, ILLEGAL_REQUEST_CODE, accountId);
    HILOG_DEBUG("%{public}s. End calling StartAbilityWithAccount. ret=%{public}d", __func__, err);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartAbilityWithAccount is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StartAbilityWithAccount(
    const AAFwk::Want &want, int32_t accountId, const AAFwk::StartOptions &startOptions) const
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    (const_cast<Want &>(want)).SetParam(START_ABILITY_TYPE, true);
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StartAbility(want, startOptions, token_,
        ILLEGAL_REQUEST_CODE, accountId);
    HILOG_DEBUG("%{public}s. End calling StartAbilityWithAccount. ret=%{public}d", __func__, err);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartAbilityWithAccount is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StartVpnExtensionAbility(const AAFwk::Want &want, int32_t accountId) const
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StartExtensionAbility(
        want, token_, accountId, AppExecFwk::ExtensionAbilityType::UNSPECIFIED);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StartVpnExtensionAbility is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::StopVpnExtensionAbility(const AAFwk::Want& want, int32_t accountId) const
{
    HILOG_DEBUG("%{public}s begin.", __func__);
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->StopExtensionAbility(
        want, token_, accountId, AppExecFwk::ExtensionAbilityType::UNSPECIFIED);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnContext::StopVpnExtensionAbility is failed %{public}d", err);
    }
    return err;
}

ErrCode VpnExtensionContext::ConnectAbilityWithAccount(
    const AAFwk::Want &want, int32_t accountId, const sptr<AbilityConnectCallback> &connectCallback) const
{
    HILOG_INFO("%{public}s begin.", __func__);
    ErrCode ret =
        ConnectionManager::GetInstance().ConnectAbilityWithAccount(token_, want, accountId, connectCallback);
    HILOG_INFO("VpnExtensionContext::ConnectAbilityWithAccount ErrorCode = %{public}d", ret);
    return ret;
}

ErrCode VpnExtensionContext::DisconnectAbility(
    const AAFwk::Want &want, const sptr<AbilityConnectCallback> &connectCallback) const
{
    HILOG_INFO("%{public}s begin.", __func__);
    ErrCode ret =
        ConnectionManager::GetInstance().DisconnectAbility(token_, want, connectCallback);
    if (ret != ERR_OK) {
        HILOG_ERROR("%{public}s end DisconnectAbility error, ret=%{public}d", __func__, ret);
    }
    HILOG_INFO("%{public}s end DisconnectAbility", __func__);
    return ret;
}

ErrCode VpnExtensionContext::TerminateAbility()
{
    HILOG_INFO("%{public}s begin.", __func__);
    ErrCode err = AAFwk::AbilityManagerClient::GetInstance()->TerminateAbility(token_, -1, nullptr);
    if (err != ERR_OK) {
        HILOG_ERROR("VpnExtensionContext::TerminateAbility is failed %{public}d", err);
    }
    HILOG_INFO("%{public}s end.", __func__);
    return err;
}

AppExecFwk::AbilityType VpnExtensionContext::GetAbilityInfoType() const
{
    std::shared_ptr<AppExecFwk::AbilityInfo> info = GetAbilityInfo();
    if (info == nullptr) {
        HILOG_ERROR("VpnContext::GetAbilityInfoType info == nullptr");
        return AppExecFwk::AbilityType::UNKNOWN;
    }

    return info->type;
}
}  // namespace AbilityRuntime
}  // namespace OHOS

File: frameworks/js/napi/vpn_extension_context/vpn_extension_context.js
/*
 * Copyright (c) 2023 Huawei Device Co., Ltd.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

let ExtensionContext = requireNapi('application.ExtensionContext');
let Caller = requireNapi('application.Caller');

const ERROR_CODE_INVALID_PARAM = 401;
const ERROR_MSG_INVALID_PARAM = 'Invalid input parameter.';
class ParamError extends Error {
  constructor() {
    super(ERROR_MSG_INVALID_PARAM);
    this.code = ERROR_CODE_INVALID_PARAM;
  }
}

class VpnExtensionContext extends ExtensionContext {
  constructor(obj) {
    super(obj);
  }

  startAbility(want, options, callback) {
    console.log('startAbility');
    return this.__context_impl__.startAbility(want, options, callback);
  }

  startAbilityAsCaller(want, options, callback) {
    console.log('startAbilityAsCaller');
    return this.__context_impl__.startAbilityAsCaller(want, options, callback);
  }

  startRecentAbility(want, options, callback) {
    console.log('startRecentAbility');
    return this.__context_impl__.startRecentAbility(want, options, callback);
  }

  connectAbility(want, options) {
    console.log('connectAbility');
    return this.__context_impl__.connectAbility(want, options);
  }

  connectVpnExtensionAbility(want, options) {
    console.log('connectVpnExtensionAbility');
    return this.__context_impl__.connectVpnExtensionAbility(want, options);
  }

  startAbilityWithAccount(want, accountId, options, callback) {
    console.log('startAbilityWithAccount');
    return this.__context_impl__.startAbilityWithAccount(want, accountId, options, callback);
  }

  startVpnExtensionAbility(want, callback) {
    console.log('startVpnExtensionAbility');
    return this.__context_impl__.startVpnExtensionAbility(want, callback);
  }

  startVpnExtensionAbilityWithAccount(want, accountId, callback) {
    console.log('startVpnExtensionAbilityWithAccount');
    return this.__context_impl__.startVpnExtensionAbilityWithAccount(want, accountId, callback);
  }

  stopVpnExtensionAbility(want, callback) {
    console.log('stopVpnExtensionAbility');
    return this.__context_impl__.stopVpnExtensionAbility(want, callback);
  }

  stopVpnExtensionAbilityWithAccount(want, accountId, callback) {
    console.log('stopVpnExtensionAbilityWithAccount');
    return this.__context_impl__.stopVpnExtensionAbilityWithAccount(want, accountId, callback);
  }

  connectAbilityWithAccount(want, accountId, options) {
    console.log('connectAbilityWithAccount');
    return this.__context_impl__.connectAbilityWithAccount(want, accountId, options);
  }

  connectVpnExtensionAbilityWithAccount(want, accountId, options) {
    console.log('connectVpnExtensionAbilityWithAccount');
    return this.__context_impl__.connectVpnExtensionAbilityWithAccount(want, accountId, options);
  }

  disconnectAbility(connection, callback) {
    console.log('disconnectAbility');
    return this.__context_impl__.disconnectAbility(connection, callback);
  }

  stopVpnExtensionAbility(connection, callback) {
    console.log('stopVpnExtensionAbility');
    return this.__context_impl__.stopVpnExtensionAbility(connection, callback);
  }

  terminateSelf(callback) {
    console.log('terminateSelf');
    return this.__context_impl__.terminateSelf(callback);
  }

  startAbilityByCall(want) {
    return new Promise(async (resolve, reject) => {
      if (typeof want !== 'object' || want == null) {
        console.log('VpnExtensionContext::startAbilityByCall input param error');
        reject(new ParamError());
        return;
      }

      let callee = null;
      try {
        callee = await this.__context_impl__.startAbilityByCall(want);
      } catch (error) {
        console.log('VpnExtensionContext::startAbilityByCall Obtain remoteObject failed');
        reject(error);
        return;
      }

      resolve(new Caller(callee));
      console.log('VpnExtensionContext::startAbilityByCall success');
      return;
    });
  }

  startAbilityByCallWithAccount(want, accountId) {
    return new Promise(async (resolve, reject) => {
      if (typeof want !== 'object' || want == null || typeof accountId !== 'number') {
        console.log('VpnExtensionContext::startAbilityByCall With accountId input param error');
        reject(new ParamError());
        return;
      }

      let callee = null;
      try {
        callee = await this.__context_impl__.startAbilityByCall(want, accountId);
      } catch (error) {
        console.log('VpnExtensionContext::startAbilityByCall With accountId Obtain remoteObject failed');
        reject(error);
        return;
      }

      resolve(new Caller(callee));
      console.log('VpnExtensionContext::startAbilityByCall With accountId success');
      return;
    });
  }
}

export default VpnExtensionContext;
